pkg/sql/sem/tree/type_check.go

```diff
@@ -546,6 +546,25 @@ func (expr *CollateExpr) TypeCheck(ctx *SemaContext, desired *types.T) (TypedExp
 		"incompatible type for COLLATE: %s", t)
 }
 
+// AdjustTypeForCollate returns the collated version of the input type.
+func AdjustTypeForCollate(typ *types.T, locale string) (*types.T, error) {
+	switch typ.Family() {
+	case types.StringFamily:
+		return types.MakeCollatedString(typ, locale), nil
+	case types.CollatedStringFamily:
+		return nil, pgerror.Newf(pgcode.Syntax, "multiple COLLATE declarations")
```

Comment:
super nit: `multiple COLLATE clauses not allowed` is the psql error, maybe we should do the same :D